### PR TITLE
Implement internal toolkit navigation

### LINKED ISSUE #318: [Toolkits] Internal links/navigation
Add support for internal links between toolkit sections following the same approach to navigation as we use in Reference. We should support the same scroll and highlight approach and share code between them.

### PR DESCRIPTION
**Pull request by microbit-robert**
Links to the internal 'explore' section mimic the 'reference' section, including scrolling and highlighting.

Closes #318

### CODE REVIEW

File: src/documentation/ToolkitDocumentation/ToolkitTopicEntry.tsx [hunk 1]
```diff
@@ -46,59 +56,114 @@ const ToolkitTopicEntry = ({
...
+          }, 3000);
+        }, 300);
+      }, 150);
+    } else if (previousAnchor !== anchor) {
+      if (ref.current && scrollable.current) {
+        scrollable.current.scrollTo({
+          top: 0,
+          behavior: prefersReducedMotion ? "auto" : "smooth",
+        });
+      }
```
Comment (microbit-robert) [inline src/documentation/ToolkitDocumentation/ToolkitTopicEntry.tsx:91]: It feels like this shouldn't be necessary, but it scrolls to the top of the toolkit area where a topic is linked, but not a specific entry.

File: src/documentation/ToolkitDocumentation/ToolkitContent.tsx [hunk 2]
```diff
@@ -75,8 +75,10 @@ const ToolkitInternalLinkMark = (
         e.preventDefault();
         setState({
           ...state,
-          // Hmm, we need to know the tab/toolkit (we should name them the same).
-          // We also need to switch to router-based navigation for the other toolkits.
+          tab: "explore",
+          explore: {
+            id: props.mark.slug.current,
```
Comment (microbit-robert) [inline src/documentation/ToolkitDocumentation/ToolkitContent.tsx:80]: We need to conditionally use the slug as it is if `targetType === "toolkitTopic"`. Else, we need to somehow get the parent topic slug and join this with the entry slug.

**Comment by microbit-robert** [comment]
Review link:
https://review-python-editor-next.microbit.org/internal-links/

File: src/documentation/ToolkitDocumentation/ToolkitTopicEntry.tsx (hunk 1 above)
Comment (microbit-robert) [reply src/documentation/ToolkitDocumentation/ToolkitTopicEntry.tsx]: resolved with commit 10d7bb5

File: src/documentation/ToolkitDocumentation/ToolkitContent.tsx (hunk 2 above)
Comment (microbit-robert) [reply src/documentation/ToolkitDocumentation/ToolkitContent.tsx:80]: Resolved with commit 87dbf74.

File: src/documentation/ToolkitDocumentation/ToolkitDocumentation.tsx [hunk 3]
```diff
@@ -60,28 +60,45 @@ const parseAnchor = (anchor: Anchor | undefined): ToolkitNavigationState => {
...
+    return parentTopic || undefined;
+  });
+
   const activeItem =
     topic && state.itemId
-      ? topic.contents?.find((i) => i.name === state.itemId)
+      ? topic.contents?.find((i) => i.slug.current === state.itemId)
+      : topic && !state.itemId
+      ? topic.contents?.find((i) => i.slug.current === state.topicId)
       : undefined;
```
Comment (microbit-robert) [inline src/documentation/ToolkitDocumentation/ToolkitDocumentation.tsx:94]: There may be a better approach to this. Currently handles routes in the form of `?tab=explore&explore=scope-changing-global-variables-in-functions` and `?tab=explore&explore=functions/scope-changing-global-variables-in-functions` which may be useful for external links pointing at the editor.

Comment (microbit-robert) [reply src/documentation/ToolkitDocumentation/ToolkitDocumentation.tsx]: Refactored in commit c02878d

File: src/documentation/ToolkitDocumentation/api.ts [hunk 4]
```diff
@@ -62,3 +100,18 @@ export const fetchToolkit = async (languageId: string): Promise<Toolkit> => {
...
+  let activeItem: ToolkitTopicEntry | undefined;
+  const topic: ToolkitTopic | undefined = toolkit.contents?.find((t) => {
+    if (t.slug.current === topicId) {
+      return t;
+    }
+    activeItem = t.contents?.find((e) => e.slug.current === topicId);
+    return activeItem?.parent;
+  });
+  return [topic, activeItem];
+};
```
Comment (microbit-robert) [inline src/documentation/ToolkitDocumentation/api.ts:117]: Is there a better way to handle TypeScript here?

File: src/documentation/ToolkitDocumentation/ToolkitDocumentation.tsx [hunk 5]
```diff
@@ -60,28 +61,27 @@ const parseAnchor = (anchor: Anchor | undefined): ToolkitNavigationState => {
...
-    topic && state.itemId
-      ? topic.contents?.find((i) => i.name === state.itemId)
-      : undefined;
+  const [topic, activeItem] = getTopicAndActiveItem(toolkit, state.topicId);
 
   if (topic) {
     return (
       <ToolkitLevel
+        // key prop used to ensure scroll position top
+        // after using internal link to toolkit topic
```
Comment (microbit-matt-hillsdon) [inline src/documentation/ToolkitDocumentation/ToolkitDocumentation.tsx:83]:
```suggestion
        // Key prop used to ensure scroll position top
        // after using internal link to toolkit topic.
```

File: src/documentation/ToolkitDocumentation/api.ts [hunk 6]
```diff
@@ -15,9 +15,40 @@ const toolkitQuery = (languageId: string): string => {
...
+        detailContent[] {
+          ...,
+          markDefs[]{
+            ...,
+            _type == "toolkitInternalLink" => {
+              "slug": @.reference->slug,
+              "targetType": @.reference->_type
+            }
+          }
+        },
```
Comment (microbit-matt-hillsdon) [inline src/documentation/ToolkitDocumentation/api.ts:51]: Good spot!

File: src/documentation/ToolkitDocumentation/api.ts [hunk 7]
```diff
@@ -46,7 +77,14 @@ const fetchToolkitInternal = async (
     if (toolkits.length > 1) {
       throw new Error("Unexpected results");
     }
-    return toolkits[0];
+    // add topic entry parent for toolkit navigation
```
Comment (microbit-matt-hillsdon) [inline src/documentation/ToolkitDocumentation/api.ts:80]:
```suggestion
    // Add topic entry parent for toolkit navigation.
```

File: src/documentation/reference/ReferenceToolkit.tsx [hunk 8]
```diff
@@ -74,7 +74,7 @@ const ActiveToolkitLevel = ({
         <List flex="1 1 auto">
           {(module.children ?? []).map((child) => (
             <ListItem key={child.id}>
-              <ReferenceNode docs={child} width="100%" anchor={anchor} mb={0} />
```
Comment (microbit-matt-hillsdon) [inline src/documentation/reference/ReferenceToolkit.tsx:77]: Matt to check this when reviewing the UI.

I remember needing to add this as reference nodes space themselves out from each other but that shouldn't apply at the top level (only e.g. for fields/methods in a class).

File: src/documentation/ToolkitDocumentation/api.ts (hunk 7 above)
Comment (microbit-robert) [reply src/documentation/ToolkitDocumentation/api.ts]: [`562fb0d`](https://github.com/microbit-foundation/python-editor-next/pull/427/commits/562fb0d3caf6afca433d729f4ee934b0c2632427)

File: src/documentation/ToolkitDocumentation/ToolkitDocumentation.tsx (hunk 5 above)
Comment (microbit-robert) [reply src/documentation/ToolkitDocumentation/ToolkitDocumentation.tsx]: [`562fb0d`](https://github.com/microbit-foundation/python-editor-next/pull/427/commits/562fb0d3caf6afca433d729f4ee934b0c2632427)

File: src/documentation/ToolkitDocumentation/Highlight.tsx [hunk 9]
```diff
@@ -1,18 +1,81 @@
...
+import { Anchor } from "../../router-hooks";
+import { useLogging } from "../../logging/logging-hooks";
+import { useScrollablePanelAncestor } from "../../workbench/ScrollablePanel";
 
 interface HighlightProps extends BoxProps {
-  value: boolean;
+  anchor?: Anchor;
+  entryName: string;
+  active: Boolean | undefined;
+  disclosure: any;
```
Comment (microbit-matt-hillsdon) [inline src/documentation/ToolkitDocumentation/Highlight.tsx:12]: To review

File: src/documentation/ToolkitDocumentation/api.ts (hunk 4 above)
Comment (microbit-matt-hillsdon) [reply src/documentation/ToolkitDocumentation/api.ts]: We made it... different.

File: src/documentation/reference/ReferenceToolkit.tsx (hunk 8 above)
Comment (microbit-matt-hillsdon) [reply src/documentation/reference/ReferenceToolkit.tsx]: We added this back.

File: src/documentation/ToolkitDocumentation/Highlight.tsx (hunk 9 above)
Comment (microbit-matt-hillsdon) [reply src/documentation/ToolkitDocumentation/Highlight.tsx]: Fixed.

**Review by microbit-matt-hillsdon — APPROVED** [review]

**Pull request merged**